aggregator/src/aggregator.rs

```diff
@@ -478,26 +564,184 @@ impl<C: Clock> Aggregator<C> {
+                })?;
+
+        let our_role = match peer_role {
+            Role::Leader => Role::Helper,
+            Role::Helper => Role::Leader,
+            _ => {
+                return Err(Error::Internal(
+                    "role should have only been Helper or Leader".to_string(),
+                ))
+            }
+        };
+
+        let vdaf_verify_keys =
+            Vec::from([peer_aggregator.derive_vdaf_verify_key(task_id, &vdaf_instance)]);
+
+        let task = taskprov::Task::new(
+            *task_id,
+            aggregator_urls,
+            task_config.query_config().query().try_into()?,
+            vdaf_instance,
+            our_role,
+            vdaf_verify_keys,
+            task_config.query_config().max_batch_query_count() as u64,
+            Some(*task_config.task_expiration()),
+            peer_aggregator.report_expiry_age().cloned(),
+            task_config.query_config().min_batch_size() as u64,
+            *task_config.query_config().time_precision(),
+            *peer_aggregator.tolerable_clock_skew(),
+        )
+        .map_err(|err| Error::InvalidTask(*task_id, OptOutReason::TaskParameters(err)))?;
+        self.datastore
+            .run_tx_with_name("taskprov_put_task", |tx| {
+                let task = task.clone();
+                Box::pin(async move { tx.put_task(&task.into()).await })
+            })
+            .await
+            .or_else(|error| -> Result<(), Error> {
+                match error {
+                    // If the task is already in the datastore, then some other request or aggregator
+                    // replica beat us to inserting it. They _should_ have inserted all the same parameters
+                    // as we would have, so we can proceed as normal.
+                    DatastoreError::MutationTargetAlreadyExists => {
+                        warn!(
+                            ?task_id,
+                            ?error,
+                            "taskprov: went to insert task into db, but it already exists"
+                        );
+                        Ok(())
+                    }
+                    error => Err(error.into()),
+                }
+            })?;
+
+        info!(?task, ?peer_aggregator, "taskprov: opted into new task");
+        Ok(())
+    }
+
+    /// Validate and authorize a taskprov request. Returns values necessary for determining whether
+    /// we can opt into the task. This function might return an opt-out error for conditions that
+    /// are relevant for all DAP workflows (e.g. task expiration).
+    #[tracing::instrument(skip(self, aggregator_auth_token), err)]
+    async fn taskprov_authorize_request(
+        &self,
+        peer_role: &Role,
+        task_id: &TaskId,
+        taskprov_header: &[u8],
+        aggregator_auth_token: Option<&AuthenticationToken>,
+    ) -> Result<(TaskConfig, Arc<PeerAggregator>, Vec<Url>), Error> {
```

Comment:
I'm very concerned about parsing the `TaskConfig` object before we can even authenticate taskprov requests. While using Rust does mean parsing untrusted inputs is safer, it still should be avoided as much as possible (I acknowledge that we have already parsed a whole lot of untrusted HTTP before getting to this point). That was a lot of the reason that we refactored the DAP API to put the task ID and things like aggregation job IDs into request paths.

By this point, we already have the `aggregator_auth_token`. Could we instead scan the `taskprov_aggregator_auth_tokens` table looking for a match? Alternately, we could augment the schema for `taskprov_peer_aggregators` so that it includes some unique peer aggregator ID and then require incoming requests to have a `dap-taskprov-peer-aggregator-id` header in them.

Reply:
> Could we instead scan the taskprov_aggregator_auth_tokens table looking for a match?

I think this can be done. It would be costly with lots of peer aggregators and/or auth tokens, but we don't plan on having that many anyways.

Reply:
Discussed in https://isrg.slack.com/archives/C0167LT4C73/p1691514380751889. This requires some additional discussion with our peers about how authentication should work.

I will leave it as-is with a FIXME comment and an issue for follow up.